File: src/main/java/com/sup/dev/java_pc/sql/SqlQueryUpdate.java
package com.sup.dev.java_pc.sql;

import java.util.ArrayList;

public class SqlQueryUpdate extends SqlQueryWithWhere {

    private String table;
    private ArrayList<UpdateColumn> columns = new ArrayList<>();

    public SqlQueryUpdate(String table, Object... columns) {
        this.table = table;
        for (int i = 0; i < columns.length; i++) {
            if (columns[i] instanceof UpdateColumn)
                addUpdate((UpdateColumn) columns[i]);
            else if (columns[i] instanceof String)
                addUpdate(new UpdateColumnSimple((String) columns[i]));
            else throw new RuntimeException("Bad field type [" + columns[i] + "]");
        }
    }

    public SqlQueryUpdate addUpdate(UpdateColumn updateColumn) {
        columns.add(updateColumn);
        return this;
    }

    public SqlQueryUpdate addUpdate(String column) {
        columns.add(new UpdateColumnSimple(column));
        return this;
    }

    public SqlQueryUpdate addUpdate(String column, Object values) {
        columns.add(new UpdateColumnSimple(column, values));
        return this;
    }

    public int getUpdateColumnsCount(){
        return columns.size();
    }

    @Override
    protected String createQuery() {
        StringBuilder sql = new StringBuilder(UPDATE + table + SET + columns.get(0).toQuery());
        for (int i = 1; i < columns.size(); i++)
            sql.append(",").append(columns.get(i).toQuery());
        sql.append(getWhereString());
        return sql.toString();
    }

    public static abstract class UpdateColumn {

        public abstract String toQuery();
    }


    public static class UpdateColumnSimple extends UpdateColumn {

        public final String column;
        public final Object values;

        public UpdateColumnSimple(String column, Object values) {
            this.column = column;
            this.values = values;
        }

        public UpdateColumnSimple(String column) {
            this(column, "?");
        }

        @Override
        public String toQuery() {
            return column + "=" + values;
        }
    }

    public static class UpdateColumnCount extends UpdateColumn {

        public final String column;
        public final String operation;
        public final Object value;

        public UpdateColumnCount(String column) {
            this(column, "?");
        }

        public UpdateColumnCount(String column, Object value) {
            this(column, "+", value);
        }

        public UpdateColumnCount(String column, String operation, Object value) {
            this.column = column;
            this.operation = operation;
            this.value = value;
        }

        @Override
        public String toQuery() {
            return column + "=" + column + operation + "(" + value + ")";
        }

    }

}

File: src/main/java/com/sup/dev/java_pc/sql/SqlQueryUpdate.kt
package com.sup.dev.java_pc.sql

import java.util.ArrayList


class SqlQueryUpdate(private val table: String, vararg columns: Any) : SqlQueryWithWhere() {
    private val columns = ArrayList<UpdateColumn>()

    val updateColumnsCount: Int
        get() = columns.size

    init {
        for (i in columns.indices) {
            if (columns[i] is UpdateColumn)
                addUpdate(columns[i] as UpdateColumn)
            else if (columns[i] is String)
                addUpdate(UpdateColumnSimple(columns[i] as String))
            else
                throw RuntimeException("Bad field type [" + columns[i] + "]")
        }
    }

    fun addUpdate(updateColumn: UpdateColumn): SqlQueryUpdate {
        columns.add(updateColumn)
        return this
    }

    fun addUpdate(column: String): SqlQueryUpdate {
        columns.add(UpdateColumnSimple(column))
        return this
    }

    fun addUpdate(column: String, values: Any): SqlQueryUpdate {
        columns.add(UpdateColumnSimple(column, values))
        return this
    }

    override fun createQuery(): String {
        val sql = StringBuilder(Sql.UPDATE + table + Sql.SET + columns[0].toQuery())
        for (i in 1 until columns.size)
            sql.append(",").append(columns[i].toQuery())
        sql.append(createWhere())
        return sql.toString()
    }

    abstract class UpdateColumn {

        abstract fun toQuery(): String
    }


    class UpdateColumnSimple @JvmOverloads constructor(val column: String, val value: Any = "?") : UpdateColumn() {

        override fun toQuery(): String {
            return "$column=$value"
        }
    }

    class UpdateColumnCount(val column: String, val operation: String, val value: Any) : UpdateColumn() {

        @JvmOverloads constructor(column: String, value: Any = "?") : this(column, "+", value) {}

        override fun toQuery(): String {
            return "$column=$column$operation($value)"
        }

    }

}
